Library/OcVirtualFsLib/VirtualDir.c

```diff
@@ -0,0 +1,586 @@
+  )
+{
+  EFI_STATUS         Status;
+  VIRTUAL_DIR_DATA   *Data;
+
+  ASSERT (This != NULL);
+
+  Data = VIRTUAL_DIR_FROM_PROTOCOL (This);
+
+  if (Data->UnderlyingProtocol != NULL) {
+    Status = Data->UnderlyingProtocol->Open (
+      Data->UnderlyingProtocol,
+      NewHandle,
+      FileName,
+      OpenMode,
+      Attributes
+      );
+    if (!EFI_ERROR (Status)) {
+      return CreateRealFile (*NewHandle, NULL, TRUE, NewHandle);
+    }
+    return Status;
+  }
+
+  //
+  // Cannot currently handle any virtualized files in directory.
+  //
+  return EFI_NOT_FOUND;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirClose (
+  IN EFI_FILE_PROTOCOL  *This
+  )
+{
+  EFI_STATUS         Status;
+  VIRTUAL_DIR_DATA   *Data;
+
+  ASSERT (This != NULL);
+
+  Data = VIRTUAL_DIR_FROM_PROTOCOL (This);
+
+  if (Data->UnderlyingProtocol != NULL) {
+    Status = Data->UnderlyingProtocol->Close (
+      Data->UnderlyingProtocol
+    );
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+  }
+
+  VirtualDirFree (This);
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirDelete (
+  IN EFI_FILE_PROTOCOL  *This
+  )
+{
+  EFI_STATUS         Status;
+
+  Status = VirtualDirClose (This);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  //
+  // Virtual directories cannot be deleted.
+  //
+  return EFI_WARN_DELETE_FAILURE;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirRead (
+  IN EFI_FILE_PROTOCOL        *This,
+  IN OUT UINTN                *BufferSize,
+     OUT VOID                 *Buffer
+  )
+{
+  EFI_STATUS            Status;
+  VIRTUAL_DIR_DATA      *Data;
+  LIST_ENTRY            *DirEntryLink;
+  VIRTUAL_DIR_ENTRY     *DirEntry;
+  UINTN                 ReadSize;
+  UINTN                 FileStrSize;
+
+  ASSERT (This != NULL);
+  ASSERT (BufferSize != NULL);
+
+  Data = VIRTUAL_DIR_FROM_PROTOCOL (This);
+
+  //
+  // If our entry position is zero, try to read underlying protocol first.
+  //
+  ReadSize = *BufferSize;
+  if (Data->CurrentEntry == NULL && Data->UnderlyingProtocol != NULL) {
+    Status = Data->UnderlyingProtocol->Read (
+      Data->UnderlyingProtocol,
+      BufferSize,
+      Buffer
+      );
+
+      if (EFI_ERROR (Status) || *BufferSize != 0) {
+        return Status;
+      }
+  }
+
+  //
+  // Ensure we have any entries at all.
+  //
+  if (IsListEmpty (&Data->Entries)) {
+    *BufferSize = 0;
+    return EFI_SUCCESS;
+  }
+
+  //
+  // End of directory.
+  //
+  if (Data->CurrentEntry != NULL && IsNodeAtEnd (&Data->Entries, Data->CurrentEntry)) {
+    Data->CurrentEntry = NULL;
+
+    *BufferSize = 0;
+    return EFI_SUCCESS;
+  }
+
+  *BufferSize = ReadSize;
+  DirEntryLink = Data->CurrentEntry != NULL ? GetNextNode (&Data->Entries, Data->CurrentEntry) : GetFirstNode (&Data->Entries);
+
+  //
+  // Get next file info structure.
+  //
+  DirEntry = GET_VIRTUAL_DIR_ENTRY_FROM_LINK (DirEntryLink);
+  if (DirEntry->FileInfo->Size < SIZE_OF_EFI_FILE_INFO) {
+    return EFI_DEVICE_ERROR;
```

Comment:
Does that mean a memory corruption as I do not see how it could happen in normal functioning. If so, please add an `ASSERT (FALSE)` here to detect it early.